arrow/parser.py

```diff
@@ -292,11 +293,16 @@ def _generate_pattern_re(self, fmt):
         # and time string in a natural language sentence. Therefore, searching
         # for a string of the form YYYY-MM-DD in "blah 1998-09-12 blah" will
         # work properly.
-        # Reference: https://stackoverflow.com/q/14232931/3820660
-        starting_word_boundary = r"(?<![\S])"
-        ending_word_boundary = r"(?![\S])"
+        # Certain punctuation before or after the target pattern such as
+        # "1998-09-12," is permitted. For the full list of valid punctuation,
+        # see the documentation.
+
+        starting_punctuation_bound = (
+            r"(?<!\S\S)(?<![^,.;:?!\"'`\[\]{}(" r")<>\s])(\b|^)"
+        )
+        ending_punctuation_bound = r"(?=[,.;:?!\"'`\[\]{}()<>]?(?!\S))"
```

Comment:
It would be great if you could provide a [regex101](regex101.com/) example to walk us through the updated regex.

Also, you should escape the following characters if you are matching them literally in a regex: `. ^ $ * + - ? ( ) [ ] { } \ |`. We escape all symbols just to be safe (add a back slash like `\{`).

Reply:
@jadchaar All great feedback. I'll wait a day or two to implement the changes to see if @systemcatch has anything to add. I'll also provide a regex101 example in the next couple days, but I can explain the regex right now:

> (?<!\S\S)

Don't have two consecutive whitespace characters. This ensures that we allow cases like `11.25.2019.` but not `1.11.25.2019` (for pattern MM.DD.YYYY)


> (?<![^,.;:?!\"'`\[\]{}()<>\s])

This is the list of punctuation that is ok before the pattern. (ie "It can't not be these characters before the pattern")


> (\b|^)

The `\b` is to block cases like `1201912` but allow `201912` for pattern YYYYMM. The `^` was necessary to allow a negative number through i.e. before epoch numbers (a strange edge case).


> (?=[,.;:?!\"'`\[\]{}()<>]?

Positive lookahead stating that these punctuation can appear after the pattern zero or one times. The `?` quantifier can't be used before the pattern due to the way regex compiles or something.


> (?!\S)

Don't allow any non-whitespace character after the punctuation.

Reply:
> > (?!\S)
> 
> Don't allow any non-whitespace character after the punctuation.

We'd probably want cases like this to pass: `arrow.get("Monday, 9. September 2019,testing", "dddd, D. MMMM YYYY")` since the punctuation acts as a separator. The original purpose of the word boundary was to simply prevent cases like `arrow.get("Monday, 9. September 2019testing", "dddd, D. MMMM YYYY")`, but since we are allowing punctuation to follow, I am not sure if that part of the regex is necessary. Open to your thoughts though.

Reply:
Hey @andrewchouman I really like your explanation for the new regex, maybe we could incorporate that into the code by using https://docs.python.org/3/library/re.html#re.VERBOSE?

Reply:
Hey @andrewchouman, disregard my previous comment. Chris and I think your assumption of not allowing non-whitespace after the punctuation is correct :).